can_validator/MY17_Can_Library_c.py

```diff
@@ -0,0 +1,168 @@
+    :param spec_path: CAN spec path
+    :param base_path: File with template code that's not autogenerated
+    :param special_cases_path: File with code for special cases of functions that aren't based off of CAN spec
+    :param unused_messages: list of messages that are in the CAN spec but unused
+    """
+    spec = ParseCAN.spec.can(spec_path)
+    with open(output_path, 'w') as f:
+        # Copy over base
+        with open(base_path) as base:
+            lines = base.readlines()
+            f.writelines(lines)
+
+        # Write switch statement
+        f.write(
+            "Can_MsgID_T Can_MsgType(void) {\n" +
+            "  lastError = Can_RawRead(&lastMessage);\n" +
+            "  if (lastError == Can_Error_NO_RX) {\n" +
+            "    return Can_No_Msg;\n" +
+            "  } else if (lastError != Can_Error_NONE) {\n" +
+            "    return Can_Error_Msg;\n" +
+            "  }\n\n" +
+            "  uint16_t id = lastMessage.id;\n" +
+            "  uint16_t first_byte = lastMessage.data[0];\n\n" +
+            "  switch(id) {\n")
+        for message in spec.messages.values():
+            if message.name in unused_messages:
+                continue
+            f.write(
+                "    case " + message.name.upper() + "__id:\n" +
+                "      return Can_" + message.name + "_Msg;\n")
+        # Special cases
+        f.write(
+            "    case MC_RESPONSE__id:\n" +
+            "      if (first_byte == CAN_MC_REG_ERRORS_AND_WARNINGS) {\n" +
+            "        return Can_MC_ErrorAndWarning_Msg;\n" +
+            "      } else if (first_byte == CAN_MC_REG_STATE) {\n" +
+            "        return Can_MC_State_Msg;\n" +
+            "      } else {\n" +
+            "        return Can_MC_DataReading_Msg;\n" +
+            "      }\n")
+        f.write(
+            "    case VCU_MC_MESSAGE__id:\n" +
+            "      if (first_byte == CAN_MC_REG_MSG_REQUEST || first_byte == CAN_MC_REG_MSG_EVENT_REQUEST) {\n" +
+            "        return Can_Vcu_MCRequest_Msg;\n" +
+            "      } else if (first_byte == CAN_MC_REG_TORQUE_CMD) {\n" +
+            "        return Can_Vcu_MCTorque_Msg;\n" +
+            "      } else {\n" +
+            "        return Can_Unknown_Msg;\n" +
+            "      }\n")
+
+        # Finish up
+        f.write(
+            "    default:\n" +
+            "      return Can_Unknown_Msg;\n" +
+            "   }\n" +
+            "}\n\n")
+
+        for message in spec.messages.values():
+            if message.name in unused_messages:
+                continue
+
+            # Write TO_CAN
+            f.write(
+                "TO_CAN(Can_" + message.name + ") {\n" +
+                "  uint64_t bitstring = 0;\n")
+
+            length = 0
+            for segment_name, segment in message.segments.items():
+                field_name = segment_name
+
+                if field_name == "unused" or field_name == "reserved":
```

Comment:
eventually you will want to define this + other spec constants in some other file and just reference those variables.

Reply:
@alec-heif I believe that anything that has to do with lack of information (eq. unused) should not belong in the spec, and it should be assumable that in those undefined ranges you can only find zeros.

Reply:
Which probably means that these constants can either be moved to a python module or the CAN library code should know how to deal with empty segments.

Reply:
For now, I can just move them into `common.py`, like with the unused messages